### PR TITLE
feat(docs): adds ability to delete docs from ReadMe if they're no longer in local folder

### PR DESCRIPTION
**Pull request by shaiarmis**
## 🧰 Changes

The `rdme docs` command syncs the documents between a local folder and a ReadMe project.
It works really well for new file in the folder, or for editing existing ones, but it does't work for files that were deleted and are no longer needed.
With the new `--deleteMissing` option (off by default) this is now possible.

## 🧬 QA & Testing

1. Sync some files from a folder with your project
2. Delete one of the files from the folder
3. Sync again (with `--deleteMissing`)
4. The file should be delete from ReadMe

### CODE REVIEW

File: src/lib/fetch.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ import APIError from './apiError';
 import isGHA from './isGitHub';
 import { debug } from './logger';
 
+const SUCCESS_NO_CONTENT = 204;
```
Comment (shaiarmis) [inline src/lib/fetch.ts:12]: The `DELETE /api/v1/docs/{slug}` endpoint returns 204 with an empty body if the deletion was successful.

File: src/lib/fetch.ts [hunk 2]
```diff
@@ -74,6 +76,9 @@ async function handleRes(res: Response) {
   // If we receive a non-JSON response, it's likely an error.
   // Let's debug the raw response body and throw it.
   const body = await res.text();
+  if (res.status === SUCCESS_NO_CONTENT) {
```
Comment (erunion) [inline src/lib/fetch.ts:79]: Can you move this above the `res.text()` call? We don't need to call that if we're going to return immediately here with an empty object.

**Review by erunion — CHANGES_REQUESTED** [review]
one minor comment but overall this seems good to me

File: src/lib/getDocs.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  }).then(handleRes);
+}
+
+/**
+ * Retrieve the docs under all categories or under a specific one
+ *
+ * @param {String} key the project API key
+ * @param {String} selectedVersion the project version
+ * @param {String} [category] Get docs from the specified category only
+ * @returns {Promise<Array<Object>>} an array containing the docs
```
Comment (kanadgupta) [inline src/lib/getDocs.ts:26]: Could you move this type to the function definition so TS picks up on it?

File: src/lib/deleteDoc.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Delete a document from ReadMe
+ *
+ * @param {String} key the project API key
+ * @param {String} selectedVersion the project version
+ * @param {Boolean} dryRun boolean indicating dry run mode
+ * @param {String} slug The slug of the document to delete
+ * @param {String} type module within ReadMe to update (e.g. docs, changelogs, etc.)
+ * @returns {Promise<String>} a string containing the result
```
Comment (kanadgupta) [inline src/lib/deleteDoc.ts:16]: Could you move this type to the function definition so TS picks up on it?

File: src/cmds/docs/index.ts [hunk 5]
```diff
@@ -67,7 +76,17 @@ export default class DocsCommand extends Command {
 
     Command.debug(`number of files: ${files.length}`);
 
-    if (!files.length) {
+    const changes: string[] = [];
+    if (deleteMissing) {
+      const docs = await getDocs(key, selectedVersion);
```
Comment (kanadgupta) [inline src/cmds/docs/index.ts:81]: Could you add a `warn` statement above this line that warns the user about what they're about to do? Using this function:

https://github.com/readmeio/rdme/blob/0292a09eabc9e0beb0e982f6f24bed4c2f757a41/src/lib/logger.ts#L27-L32

File: src/lib/getSlug.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * @returns {String} a string containing the slug of the file
+ */
+export default function getSlug(filepath: string) {
+  debug(`reading file ${filepath}`);
+  const file = fs.readFileSync(filepath, 'utf8');
+  const matter = grayMatter(file);
+  debug(`frontmatter for ${filepath}: ${JSON.stringify(matter)}`);
+
+  // Stripping the subdirectories and markdown extension from the filename and lowercasing to get the default slug.
+  return matter.data.slug || path.basename(filepath).replace(path.extname(filepath), '').toLowerCase();
```
Comment (kanadgupta) [inline src/lib/getSlug.ts:22]: This logic is a duplication of this logic:

https://github.com/readmeio/rdme/blob/0292a09eabc9e0beb0e982f6f24bed4c2f757a41/src/lib/pushDoc.ts#L34-L41

What if this function was renamed to `readDoc`[^1] and it returned an object containing the `slug`, the `matter` data, and the `file`? Then you could reuse this function in `pushDoc`!

[^1]: `getDoc` would be confusing since it's reading a local file as opposed to making an API request

File: src/lib/getDocs.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+export default async function getDocs(key: string, selectedVersion: string, category?: string) {
+  if (category) {
+    return getCategoryDocs(key, selectedVersion, category);
+  }
+
+  return getCategories(key, selectedVersion)
+    .then(categories => categories.filter(({ type }: { type: string }) => type === 'guide'))
+    .then(categories => categories.map(({ slug }: { slug: string }) => getCategoryDocs(key, selectedVersion, slug)))
+    .then(Promise.all.bind(Promise))
+    .then(args => [].concat(...args));
```
Comment (kanadgupta) [inline src/lib/getDocs.ts:37]: These two lines are difficult to follow, any way they can be simplified?

File: src/cmds/docs/index.ts [hunk 8]
```diff
@@ -67,7 +76,17 @@ export default class DocsCommand extends Command {
...
+      const docSlugs = docs.map(({ slug }: { slug: string }) => slug);
+      const fileSlugs = new Set(files.map(getSlug));
+      const slugsToDelete = docSlugs.filter((slug: string) => !fileSlugs.has(slug));
+      const deletedDocs = await Promise.all(
+        slugsToDelete.map((slug: string) => deleteDoc(key, selectedVersion, dryRun, slug, this.cmdCategory))
+      );
+      changes.push(...deletedDocs);
+    } else if (!files.length) {
```
Comment (kanadgupta) [inline src/cmds/docs/index.ts:89]: What happens if the `files` array is empty and someone passes in the `--deleteMissing` flag? Could you move this logic below this block below this block? https://github.com/readmeio/rdme/blob/0292a09eabc9e0beb0e982f6f24bed4c2f757a41/src/cmds/docs/index.ts#L89-L91

File: src/lib/getDocs.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Retrieve the docs under all categories or under a specific one
+ *
+ * @param {String} key the project API key
+ * @param {String} selectedVersion the project version
+ * @param {String} [category] Get docs from the specified category only
+ * @returns {Promise<Array<Object>>} an array containing the docs
+ */
+export default async function getDocs(key: string, selectedVersion: string, category?: string) {
```
Comment (kanadgupta) [inline src/lib/getDocs.ts:28]: Is this `category` parameter even being used anywhere? If not, can you remove it?

File: __tests__/cmds/docs/index.test.ts [hunk 10]
```diff
@@ -374,6 +374,63 @@ describe('rdme docs', () => {
...
+        .get('/api/v1/categories?perPage=20&page=1')
+        .basicAuth({ user: key })
+        .reply(200, [{ slug: 'category1', type: 'guide' }], { 'x-total-count': '1' })
+        .get('/api/v1/categories/category1/docs')
+        .basicAuth({ user: key })
+        .reply(200, [{ slug: 'thisDocShouldBeMissingInFolder' }])
+        .delete('/api/v1/docs/thisDocShouldBeMissingInFolder')
+        .basicAuth({ user: key })
+        .reply(204, '');
+      const folderWithoutDocs = './__tests__/__fixtures__/ref-oas';
```
Comment (kanadgupta) [inline __tests__/cmds/docs/index.test.ts:393]: Wait so is the idea that it'll delete every doc in ReadMe since the local folder is completely empty? Can you add a test case with a non-empty folder?

**Review by kanadgupta — CHANGES_REQUESTED** [review]

File: src/lib/fetch.ts (hunk 2 above)
Comment (shaiarmis) [reply src/lib/fetch.ts:79]: Done

File: src/lib/getDocs.ts (hunk 3 above)
Comment (shaiarmis) [reply src/lib/getDocs.ts:26]: Done

File: src/lib/deleteDoc.ts (hunk 4 above)
Comment (shaiarmis) [reply src/lib/deleteDoc.ts:16]: Done

File: src/cmds/docs/index.ts (hunk 5 above)
Comment (shaiarmis) [reply src/cmds/docs/index.ts:81]: Done (using `Command.warn`)

File: src/lib/getSlug.ts (hunk 6 above)
Comment (shaiarmis) [reply src/lib/getSlug.ts:22]: Great idea! Done.

File: src/lib/getDocs.ts (hunk 7 above)
Comment (shaiarmis) [reply src/lib/getDocs.ts:37]: Done (I think ;) )

File: src/cmds/docs/index.ts (hunk 8 above)
Comment (shaiarmis) [reply src/cmds/docs/index.ts:89]: Actually that's on purpose. If all files were deleted from the folder then technically all files should be removed from ReadMe. But I do see how it may be a problem if someone accidentally passes a wrong empty folder.
Maybe in this case I can add a confirmation prompt? "This action will remove all files from ReadMe, are you sure you with to continue?" WDYT?

File: src/lib/getDocs.ts (hunk 9 above)
Comment (shaiarmis) [reply src/lib/getDocs.ts:28]: Removed

File: __tests__/cmds/docs/index.test.ts (hunk 10 above)
Comment (shaiarmis) [reply __tests__/cmds/docs/index.test.ts:393]: Refactored both tests to use a folder with a document that should not be deleted

File: src/cmds/docs/index.ts (hunk 8 above)
Comment (shaiarmis) [reply src/cmds/docs/index.ts]: Added the confirmation prompt for now, LMK what you think

File: README.md [hunk 11]
```diff
@@ -204,6 +204,10 @@ rdme docs path-to-markdown-files --version={project-version}
 
 This command also has a dry run mode, which can be useful for initial setup and debugging. You can read more about dry run mode [in our docs](https://docs.readme.com/docs/rdme#dry-run-mode).
 
+#### Deleting missing documents
+
+If you wish to delete documents from ReadMe that are no longer present in the directory, pass the `--deleteMissing` option to the command.
```
Comment (erunion) [inline README.md:209]: We chatted offline about this a bit and do you mind renaming this new option everywhere to `--cleanup`?

```suggestion
If you wish to delete documents from ReadMe that are no longer present in the directory, pass the `--cleanup` option to the command.
```

**Review by erunion — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (shaiarmis) [reply README.md]: Done :)

**Comment by shaiarmis** [comment]
> Thanks so much for adding this. LGTM but letting @kanadgupta have final sign off before it's merged.

With pleasure, that's what OpenSource is all about :)

Is there an expected timeline for the last two changes to be published to NPM?

**Comment by kanadgupta** [comment]
> Is there an expected timeline for the last two changes to be published to NPM?

We're prepping for a big v8 release, so it won't be for at least another week or two!

**Comment by shaiarmis** [comment]
> > Is there an expected timeline for the last two changes to be published to NPM?
> 
> We're prepping for a big v8 release, so it won't be for at least another week or two!

Week or two is great :)

File: README.md [hunk 12]
```diff
@@ -204,6 +204,10 @@ rdme docs path-to-markdown-files --version={project-version}
 
 This command also has a dry run mode, which can be useful for initial setup and debugging. You can read more about dry run mode [in our docs](https://docs.readme.com/docs/rdme#dry-run-mode).
 
+#### Cleanup
+
+If you wish to delete documents from ReadMe that are no longer present in the directory, pass the `--cleanup` option to the command.
```
Comment (kanadgupta) [inline README.md:209]:
```suggestion
If you wish to delete documents from ReadMe that are no longer present in your local directory, pass the `--cleanup` option to the command.
```

File: src/lib/prompts.ts [hunk 13]
```diff
@@ -220,3 +220,11 @@ export function createVersionPrompt(
     },
   ];
 }
+
+export function deleteDocsPrompt(version: string): PromptObject {
+  return {
+    type: 'confirm',
+    name: 'deleteAll',
+    message: `This action will delete all docs under version ${version} from ReadMe, are you sure you wish to continue?`,
```
Comment (kanadgupta) [inline src/lib/prompts.ts:228]:
```suggestion
    message: `This action will delete all docs under version ${version} from ReadMe, would you like to continue?`,
```

**Review by kanadgupta — CHANGES_REQUESTED** [review]
A few small comments below but this is looking good! Thanks @shaiarmis 🙂

File: __tests__/cmds/docs/index.test.ts [hunk 14]
```diff
@@ -380,6 +381,99 @@ describe('rdme docs', () => {
...
+        })
+      ).resolves.toBe(
+        '🎭 dry run! This will delete `this-doc-should-be-missing-in-folder`.\n' +
+          '🎭 dry run! `some-doc` will not be updated because there were no changes.'
+      );
+      apiMocks.done();
+      versionMock.done();
+    });
+
+    it('should do nothing if using --cleanup but thr folder is empty and the user aborted', async () => {
```
[the remark this answers is not in the document]
Comment (shaiarmis) [reply __tests__/cmds/docs/index.test.ts:456]: Oopsi, done

File: __tests__/cmds/docs/index.test.ts [hunk 15]
```diff
@@ -380,6 +381,99 @@ describe('rdme docs', () => {
     });
   });
 
+  describe('cleanup docs', () => {
```
[the remark this answers is not in the document]
Comment (shaiarmis) [reply __tests__/cmds/docs/index.test.ts:384]: Done

File: src/lib/deleteDoc.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    headers: cleanHeaders(
+      key,
+      new Headers({
+        'x-readme-version': selectedVersion,
+        'Content-Type': 'application/json',
+      })
+    ),
+  })
+    .then(handleRes)
+    .then(() => `successfully deleted \`${slug}\`.`);
```
[the remark this answers is not in the document]
Comment (shaiarmis) [reply src/lib/deleteDoc.ts:39]: Done

File: README.md (hunk 12 above)
Comment (shaiarmis) [reply README.md:209]: Done

File: src/lib/prompts.ts (hunk 13 above)
Comment (shaiarmis) [reply src/lib/prompts.ts:228]: Done
